controllers/metal3.io/hostfirmwarecomponents_controller.go

```diff
@@ -0,0 +1,323 @@
+	reason := reasonValidComponent
+	generation := info.hfc.GetGeneration()
+	// Log the components we have
+	info.log.Info("firmware components for node", "components", components, "bmh", info.bmh.Name)
+	if componentsInfoMismatch {
+		if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsChangeDetected, metav1.ConditionTrue, reason, "") {
+			dirty = true
+		}
+	}
+	if dirty {
+		info.log.Info("Components Status for HostFirmwareComponents changed")
+		info.hfc.Status.Components = components
+
+		t := metav1.Now()
+		info.hfc.Status.LastUpdated = &t
+		return r.Status().Update(info.ctx, info.hfc)
+	}
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *HostFirmwareComponentsReconciler) SetupWithManager(mgr ctrl.Manager, maxConcurrentReconcile int) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&metal3api.HostFirmwareComponents{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: maxConcurrentReconcile}).
+		WithEventFilter(
+			predicate.Funcs{
+				UpdateFunc: r.updateEventHandler,
+			}).
+		Complete(r)
+}
+
+func (r *HostFirmwareComponentsReconciler) updateEventHandler(e event.UpdateEvent) bool {
+	r.Log.Info("hostfirmwarecomponents in event handler")
+
+	if e.ObjectNew.GetGeneration() != e.ObjectOld.GetGeneration() {
+		r.Log.Info("returning true as generation changed from event handler")
+	}
+
+	return false
+}
+
+func (r *HostFirmwareComponentsReconciler) validateHostFirmwareComponents(info *rhfcInfo) []error {
+	var errors []error
+	allowedNames := map[string]struct{}{"bmc": {}, "bios": {}}
+	for _, update := range info.hfc.Spec.Updates {
+		componentName := update.Component
+		if _, ok := allowedNames[componentName]; !ok {
+			errors = append(errors, fmt.Errorf("component %s is invalid, only 'bmc' or 'bios' are allowed as update names", update.Component))
+		}
+	}
+
+	return errors
+}
+
+func (r *HostFirmwareComponentsReconciler) publishEvent(ctx context.Context, request ctrl.Request, event corev1.Event) {
+	reqLogger := r.Log.WithValues("hostfirmwarecomponents", request.NamespacedName)
+	reqLogger.Info("publishing event", "reason", event.Reason, "message", event.Message)
+	err := r.Create(ctx, &event)
+	if err != nil {
+		reqLogger.Info("failed to record event, ignoring",
+			"reason", event.Reason, "message", event.Message, "error", err)
+	}
+}
+
+func setUpdatesCondition(generation int64, status *metal3api.HostFirmwareComponentsStatus, info *rhfcInfo,
+	cond metal3api.UpdatesConditionType, newStatus metav1.ConditionStatus,
+	reason conditionReasonHFC, message string) bool {
+	newCondition := metav1.Condition{
+		Type:               string(cond),
+		Status:             newStatus,
+		ObservedGeneration: generation,
+		Reason:             string(reason),
+		Message:            message,
+	}
+	meta.SetStatusCondition(&status.Conditions, newCondition)
+
+	currCond := meta.FindStatusCondition(info.hfc.Status.Conditions, string(cond))
```

Comment:
You'll want to get the previous condition before setting the new condition (on line 316) otherwise they will always turn out to be equal.

Reply:
done